File: spotlight/src/main/java/com/takusemba/spotlight/Spotlight.java
package com.takusemba.spotlight;

import android.animation.Animator;
import android.animation.TimeInterpolator;
import android.app.Activity;
import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.DecelerateInterpolator;
import androidx.annotation.ColorRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.takusemba.spotlight.target.Target;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Arrays;

/**
 * Spotlight that holds all the {@link Target}s and show and hide {@link Target} properly,
 * and show and hide {@link SpotlightView} properly.
 **/
public class Spotlight {

  @ColorRes private static final int DEFAULT_OVERLAY_COLOR = R.color.background;
  private static final long DEFAULT_DURATION = 1000L;
  private static final TimeInterpolator DEFAULT_ANIMATION = new DecelerateInterpolator(2f);

  private static WeakReference<SpotlightView> spotlightViewWeakReference;
  private static WeakReference<Activity> contextWeakReference;
  private ArrayList<? extends Target> targets;
  private long duration = DEFAULT_DURATION;
  private TimeInterpolator animation = DEFAULT_ANIMATION;
  private OnSpotlightStateChangedListener spotlightListener;
  private int overlayColor = DEFAULT_OVERLAY_COLOR;
  private boolean isClosedOnTouchedOutside = true;

  private Spotlight(Activity activity) {
    contextWeakReference = new WeakReference<>(activity);
  }

  public static Spotlight with(@NonNull Activity activity) {
    return new Spotlight(activity);
  }

  private static Context getContext() {
    return contextWeakReference.get();
  }

  @Nullable private static SpotlightView getSpotlightView() {
    return spotlightViewWeakReference.get();
  }

  /**
   * sets {@link Target}s to Spotlight
   *
   * @param targets targets to show
   * @return the Spotlight
   */
  @SafeVarargs public final <T extends Target> Spotlight setTargets(@NonNull T... targets) {
    this.targets = new ArrayList<>(Arrays.asList(targets));

    return this;
  }

  /**
   * sets {@link Target}s to Spotlight
   *
   * @param targets targets as ArrayList to show
   * @return the Spotlight
   */
  public final <T extends Target> Spotlight setTargets(@NonNull ArrayList<T> targets) {
    this.targets = targets;

    return this;
  }

  /**
   * sets spotlight background color to Spotlight
   *
   * @param overlayColor background color to be used for the spotlight overlay
   * @return the Spotlight
   */
  public Spotlight setOverlayColor(@ColorRes int overlayColor) {
    this.overlayColor = overlayColor;
    return this;
  }

  /**
   * sets duration to {@link Target} Animation
   *
   * @param duration duration of Target Animation
   * @return the Spotlight
   */
  public Spotlight setDuration(long duration) {
    this.duration = duration;
    return this;
  }

  /**
   * sets interpolator to {@link Target} Animation
   *
   * @param animation type of Target Animation
   * @return the Spotlight
   */
  public Spotlight setAnimation(TimeInterpolator animation) {
    this.animation = animation;
    return this;
  }

  /**
   * Sets {@link OnSpotlightStateChangedListener}
   *
   * @param listener OnSpotlightEndedListener of Spotlight
   * @return This Spotlight
   */
  public Spotlight setOnSpotlightStateListener(
      @NonNull final OnSpotlightStateChangedListener listener) {
    spotlightListener = listener;
    return this;
  }

  /**
   * Sets if Spotlight closes Target if touched outside
   *
   * @param isClosedOnTouchedOutside OnSpotlightEndedListener of Spotlight
   * @return This Spotlight
   */
  public Spotlight setClosedOnTouchedOutside(boolean isClosedOnTouchedOutside) {
    this.isClosedOnTouchedOutside = isClosedOnTouchedOutside;
    return this;
  }

  /**
   * Shows {@link SpotlightView}
   */
  public void start() {
    spotlightView();
  }

  /**
   * close the current {@link Target}
   */
  public void closeCurrentTarget() {
    finishTarget();
  }

  /**
   * close the {@link Spotlight}
   */
  public void closeSpotlight() {
    finishSpotlight();
  }

  /**
   * Creates the spotlight view and starts
   */
  @SuppressWarnings("unchecked") private void spotlightView() {
    if (getContext() == null) {
      throw new RuntimeException("context is null");
    }
    final View decorView = ((Activity) getContext()).getWindow().getDecorView();
    SpotlightView spotlightView =
        new SpotlightView(getContext(), overlayColor, new OnSpotlightListener() {
          @Override public void onSpotlightViewClicked() {
            if (isClosedOnTouchedOutside) {
              finishTarget();
            }
          }
        });
    spotlightViewWeakReference = new WeakReference<>(spotlightView);
    ((ViewGroup) decorView).addView(spotlightView);
    startSpotlight();
  }

  /**
   * show Target
   */
  @SuppressWarnings("unchecked") private void startTarget() {
    if (targets != null && targets.size() > 0 && getSpotlightView() != null) {
      final Target target = targets.get(0);
      SpotlightView spotlightView = getSpotlightView();
      spotlightView.removeAllViews();
      spotlightView.addView(target.getOverlay());
      spotlightView.turnUp(target, new AbstractAnimatorListener() {
        @Override public void onAnimationStart(Animator animation) {
          if (target.getListener() != null) target.getListener().onStarted(target);
        }
      });
    }
  }

  /**
   * show Spotlight
   */
  private void startSpotlight() {
    if (getSpotlightView() == null) return;
    getSpotlightView().startSpotlight(duration, animation, new AbstractAnimatorListener() {
      @Override public void onAnimationStart(Animator animation) {
        if (spotlightListener != null) spotlightListener.onStarted();
      }

      @Override public void onAnimationEnd(Animator animation) {
        startTarget();
      }
    });
  }

  /**
   * hide Target
   */
  @SuppressWarnings("unchecked") private void finishTarget() {
    if (targets != null && targets.size() > 0 && getSpotlightView() != null) {
      getSpotlightView().turnDown(new AbstractAnimatorListener() {
        @Override public void onAnimationEnd(Animator animation) {
          if (!targets.isEmpty()) {
            Target target = targets.remove(0);
            if (target.getListener() != null) target.getListener().onEnded(target);
            if (targets.size() > 0) {
              startTarget();
            } else {
              finishSpotlight();
            }
          }
        }
      });
    }
  }

  /**
   * hide Spotlight
   */
  private void finishSpotlight() {
    if (getSpotlightView() == null) return;
    getSpotlightView().finishSpotlight(duration, animation, new AbstractAnimatorListener() {
      @Override public void onAnimationEnd(Animator animation) {
        Activity activity = (Activity) getContext();
        if (activity != null) {
          final View decorView = activity.getWindow().getDecorView();
          ((ViewGroup) decorView).removeView(getSpotlightView());
          if (spotlightListener != null) spotlightListener.onEnded();
        }
      }
    });
  }
}

File: spotlight/src/main/java/com/takusemba/spotlight/Spotlight.kt
package com.takusemba.spotlight

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.TimeInterpolator
import android.app.Activity
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.view.animation.DecelerateInterpolator
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.core.content.ContextCompat
import java.util.concurrent.TimeUnit

/**
 * Holds all of the [Target]s and [SpotlightView] to show/hide [Target], [SpotlightView] properly.
 * [SpotlightView] can be controlled with [start]/[finish].
 * All of the [Target]s can be controlled with [next]/[previous]/[show].
 *
 * Once you finish the current [Spotlight] with [finish], you can not start the [Spotlight] again
 * unless you create a new [Spotlight] to start again.
 */
class Spotlight private constructor(
    private val spotlight: SpotlightView,
    private var targets: Array<Target>,
    private val duration: Long,
    private val interpolator: TimeInterpolator,
    private val container: ViewGroup,
    private val spotlightListener: OnSpotlightListener?
) {

  private var state: SpotlightState = SpotlightState.STOPPED

  private var currentIndex = NO_POSITION

  private var currentTarget: Target? = null

  private fun initSpotlightOverlay() {
    container.addView(spotlight, MATCH_PARENT, MATCH_PARENT)
  }

  /**
   * Starts [SpotlightView] and show the first [Target].
   */
  fun start() {
    initSpotlightOverlay()
    startSpotlight()
  }

  /**
   * Closes the current [Target] if exists, and shows a [Target] at the specified [index].
   * If target is not found at the [index], it will throw an exception.
   */
  fun show(index: Int) {
    showTarget(index)
  }

  /**
   * Closes the current [Target] if exists, and shows the next [Target].
   * If the next [Target] is not found, Spotlight will finish.
   */
  fun next() {
    showTarget(currentIndex + 1)
  }

  /**
   * Closes the current [Target] if exists, and shows the previous [Target].
   * If the previous target is not found, it will throw an exception.
   */
  fun previous() {
    showTarget(currentIndex - 1)
  }

  /**
   * Closes Spotlight and [SpotlightView] will remove all children and be removed from the [container].
   */
  fun finish() {
    finishSpotlight()
  }

  /**
   * Adds new targets to an existing [Spotlight].
   */
  fun addTarget(target: Target) {
    targets = targets.plus(target)
  }

  /**
   * Sets [Target]s to show on [Spotlight].
   */
  fun setTargets(targets: List<Target>) {
    require(targets.isNotEmpty()) { "targets should not be empty. " }
    this.targets = targets.toTypedArray()
    if (isRunning()) show(0)
  }

  /**
   * Starts Spotlight.
   */
  private fun startSpotlight() {
    spotlight.apply {
      state = SpotlightState.RUNNING

      startSpotlight(duration, interpolator, object : AnimatorListenerAdapter() {
        override fun onAnimationStart(animation: Animator) {
          spotlightListener?.onStarted()
        }

        override fun onAnimationEnd(animation: Animator) {
          showTarget(0)
        }
      })
    }
  }

  /**
   * Closes the current [Target] if exists, and show the [Target] at [index].
   */
  private fun showTarget(index: Int) {
    if (currentIndex == NO_POSITION) {
      require(targets.isNotEmpty()) { "targets should not be empty. " }
      currentIndex = index
      currentTarget = targets[currentIndex]
          .also {
            spotlight.startTarget(it)
            it.listener?.onStarted()
          }
    } else {
      spotlight.finishTarget(object : AnimatorListenerAdapter() {
        override fun onAnimationEnd(animation: Animator) {
          val previousTarget = currentTarget
          previousTarget?.listener?.onEnded()
          if (index < targets.size) {
            val target = targets[index]
            currentIndex = index
            spotlight.startTarget(target)
            target.listener?.onStarted()
          } else {
            finishSpotlight()
          }
        }
      })
    }
  }

  /**
   * Closes Spotlight.
   */
  private fun finishSpotlight() {
    spotlight.apply {
      state = SpotlightState.STOPPED
      targets = emptyArray()
      currentTarget = null
      currentIndex = NO_POSITION

      finishSpotlight(duration, interpolator, object : AnimatorListenerAdapter() {
        override fun onAnimationEnd(animation: Animator) {
          spotlight.cleanup()
          container.removeView(spotlight)
          spotlightListener?.onEnded()
        }
      })
    }
  }

  /**
   * Convenience method for [Spotlight] state checks.
   */
  fun isRunning() = state == SpotlightState.RUNNING

  companion object {

    private const val NO_POSITION = -1
  }

  /**
   * Builder to build [Spotlight].
   * All parameters should be set in this [Builder].
   */
  class Builder(private val activity: Activity) {

    private var targets: Array<Target>? = null
    private var duration: Long = DEFAULT_DURATION
    private var interpolator: TimeInterpolator = DEFAULT_ANIMATION
    @ColorInt private var backgroundColor: Int = DEFAULT_OVERLAY_COLOR
    private var container: ViewGroup? = null
    private var listener: OnSpotlightListener? = null
    private var spotlightCoordinator: SpotlightCoordinator? = null

    /**
     * Sets [Target]s to show on [Spotlight].
     */
    fun setTargets(vararg targets: Target): Builder = apply {
      require(targets.isNotEmpty()) { "targets should not be empty. " }
      this.targets = arrayOf(*targets)
    }

    /**
     * Sets [Target]s to show on [Spotlight].
     */
    fun setTargets(targets: List<Target>): Builder = apply {
      require(targets.isNotEmpty()) { "targets should not be empty. " }
      this.targets = targets.toTypedArray()
    }

    /**
     * Sets [duration] to start/finish [Spotlight].
     */
    fun setDuration(duration: Long): Builder = apply {
      this.duration = duration
    }

    /**
     * Sets [backgroundColor] resource on [Spotlight].
     */
    fun setBackgroundColorRes(@ColorRes backgroundColorRes: Int): Builder = apply {
      this.backgroundColor = ContextCompat.getColor(activity, backgroundColorRes)
    }

    /**
     * Sets [backgroundColor] on [Spotlight].
     */
    fun setBackgroundColor(@ColorInt backgroundColor: Int): Builder = apply {
      this.backgroundColor = backgroundColor
    }

    /**
     * Sets [interpolator] to start/finish [Spotlight].
     */
    fun setAnimation(interpolator: TimeInterpolator): Builder = apply {
      this.interpolator = interpolator
    }

    /**
     * Sets [container] to hold [SpotlightView]. DecoderView will be used if not specified.
     */
    fun setContainer(container: ViewGroup) = apply {
      this.container = container
    }

    /**
     * Sets [OnSpotlightListener] to notify the state of [Spotlight].
     */
    fun setOnSpotlightListener(listener: OnSpotlightListener): Builder = apply {
      this.listener = listener
    }

    /**
     * Sets [SpotlightCoordinator] to keep reference to a [Spotlight].
     */
    fun setSpotlightCoordinator(spotlightCoordinator: SpotlightCoordinator): Builder = apply {
      this.spotlightCoordinator = spotlightCoordinator
    }

    fun build(): Spotlight {

      val spotlight = SpotlightView(activity, null, 0, backgroundColor)
      val container = container ?: activity.window.decorView as ViewGroup

      return Spotlight(
          spotlight = spotlight,
          targets = targets ?: emptyArray(),
          duration = duration,
          interpolator = interpolator,
          container = container,
          spotlightListener = listener
      )
          .also { spotlightCoordinator?.spotlight = it }
    }

    companion object {

      private val DEFAULT_DURATION = TimeUnit.SECONDS.toMillis(1)

      private val DEFAULT_ANIMATION = DecelerateInterpolator(2f)

      @ColorInt private val DEFAULT_OVERLAY_COLOR: Int = 0x6000000
    }
  }
}
